### PR TITLE
Reference implementations for Loop and TensorIterator ops

### PR DESCRIPTION
**Pull request by itikhono**
1. Implemented ref implementations for Loop, TensorIterator and ngraph::function
2. Enabled onnx importer unit tests for Loop op
3. Enabled single layer tests for Loop op
4. Implemented single layer tests for TensorIterator
Misc: fixes (Reduce 1d cases in mkl, Ceiling naming in mkl), refactoring

### CODE REVIEW

**Review by dmitry-gorokhov — APPROVED** [review]
CPU plug-in changes looks good to me

**Review by lazarevevgeny — APPROVED** [review]

File: ngraph/core/reference/src/runtime/reference/function.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+#include "ngraph/runtime/reference/concat.hpp"
+#include "ngraph/runtime/tensor.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            bool call(const std::vector<std::shared_ptr<runtime::Tensor>>& outputs,
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/function.cpp:32]: Why do you pass vectors of `runtime::Tensor` instead of vectors of `HostTensor`s?
Let's change api and avoid static casts

Comment (itikhono) [reply ngraph/core/reference/src/runtime/reference/function.cpp:32]: directly from Interpreter, will fix it

**Review by iefode — APPROVED** [review]
One more comment, it will be good to add TI single layer  tests

File: ngraph/test/util/engine/interpreter_engine.cpp [hunk 2]
```diff
@@ -116,8 +116,8 @@ test::INTERPRETER_Engine::INTERPRETER_Engine(const std::shared_ptr<Function> fun
     m_executable = m_backend->compile(m_function);
     for (auto i = 0; i < m_function->get_output_size(); ++i)
     {
-        m_result_tensors.push_back(m_backend->create_tensor(m_function->get_output_element_type(i),
-                                                            m_function->get_output_shape(i)));
+        m_result_tensors.push_back(m_backend->create_dynamic_tensor(
```
Comment (mbencer) [inline ngraph/test/util/engine/interpreter_engine.cpp:119]: @itikhono We have dedicated constructor to support dynamic tensors (https://github.com/openvinotoolkit/openvino/blob/bc346c3aeff8ce914b570d70c16e42ff7210c91c/ngraph/test/util/engine/interpreter_engine.cpp#L132). In order to enable it, `TestCase` should be created with `TestCaseType::DYNAMIC` argument (`auto test_case = test::TestCase<TestEngine, TestCaseType::DYNAMIC>(function);`)
I can do it in my PR with Loop tests enablement, but please revert this change.

**Review by mbencer — CHANGES_REQUESTED** [review]

File: ngraph/core/reference/src/runtime/reference/function.cpp (hunk 1 above)
Comment (itikhono) [reply ngraph/core/reference/src/runtime/reference/function.cpp]: fixed

File: ngraph/test/util/engine/interpreter_engine.cpp (hunk 2 above)
Comment (itikhono) [reply ngraph/test/util/engine/interpreter_engine.cpp]: fixed

File: ngraph/core/reference/include/ngraph/runtime/reference/function.hpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include <vector>
+#include "ngraph/function.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            HostTensorVector function(const std::shared_ptr<Function>& function,
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: Why do we use another syntax?
If I remember right all reference functions have next signature:  `void name(HostTensorVector in, HostTensorVector out, ...)`

Comment (itikhono) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: I tried to implement close to this code
```
file: inference-engine/tests/ngraph_functions/src/utils/ngraph_helpers.cpp
std::vector<std::vector<std::uint8_t>> interpreterFunction(const std::shared_ptr<Function> &function, const std::vector<std::vector<std::uint8_t>> &inputs,
                                                           std::vector<ngraph::element::Type_t> convertType)
```
Is it critical?

Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: No it is not critical, but will be better if we use one approach for all functions))

Comment (itikhono) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: " for all functions" Could you please give some examples? 
Do you mean ctors? Or what?

**Review by AlexPeskov — APPROVED** [review]

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/tensor_iterator.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace {
+    // output values increase rapidly without clip, so use only seq_lenghts = 2
+    std::vector<bool> should_decompose = {true, false};
+    std::vector<size_t> seq_lengths_zero_clip{2};
+    std::vector<size_t> seq_lengths_clip_non_zero{20};
+    std::vector<size_t> batch{1, 10};
+    std::vector<size_t> hidden_size{1, 10};
+    std::vector<size_t> input_size{10};
+    std::vector<TensorIteratorBody> body_type = {TensorIteratorBody::LSTM, TensorIteratorBody::RNN, TensorIteratorBody::GRU};
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/tensor_iterator.cpp:20]: I think it's convenient for now, TI is more often used for Sequences.  I will add another workload a bit later.

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/loop.hpp [hunk 5]
```diff
@@ -99,46 +98,6 @@ class TrivialLoopTest : public testing::WithParamInterface<LayerTestsUtils::basi
...
-            auto blob = generator(data->getTensorDesc());
-            auto blob_size = blob->byteSize();
-            auto blob_ptr = blob->buffer().as<uint8_t*>();
-
-            auto &res = res_collection[i];
-            res.resize(blob_size);
-            std::copy(blob_ptr, blob_ptr + blob_size, res.begin());
-        }
-        return res_collection;
-    }
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/include/single_layer_tests/loop.hpp:141]: No problem!

File: ngraph/core/reference/include/ngraph/runtime/reference/function.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: I guess we will use it in a `evaluate `method of ngraph::function class, so reference has exactly the same name.
@ilyachur @GlebKazantaev what do you think?

**Review by mbencer — APPROVED** [review]

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#include <vector>
+#include <memory>
+#include <ngraph/op/util/attr_types.hpp>
+#include "functional_test_utils/layer_test_utils.hpp"
+#include "ngraph_functions/builders.hpp"
+#include "ngraph_functions/utils/ngraph_helpers.hpp"
+
+namespace LayerTestsDefinitions {
+
+enum class TensorIteratorBody {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp:18]: Looks like, we can move this enum to [`ngraph_helpers`](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/ngraph_functions/include/ngraph_functions/utils/ngraph_helpers.hpp) (Can be reused in the future)

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    RNN,
+    GRU,
+    LSTM,
+    // CNN todo: implement
+};
+
+using TensorIteratorParams = typename std::tuple<
+        bool,                                     // using unroll tensor iterator transformation
+        size_t,                                   // seq_lengths
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp:27]: Is it aligned with [spec](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/infrastructure/TensorIterator_1.md)?

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp [hunk 8]
```diff
@@ -53,7 +53,7 @@ namespace LayerTestsDefinitions {
 
     void LoopTest::SetUp() {
         SKIP_IF_CURRENT_TEST_IS_DISABLED()
-        SetRefMode(LayerTestsUtils::IE);
+        SetRefMode(LayerTestsUtils::INTERPRETER);
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp:56]: I think, we can remove this line (It's default option)

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp [hunk 9]
```diff
@@ -161,7 +161,7 @@ namespace LayerTestsDefinitions {
 
     void StaticShapeLoopTest::SetUp() {
         SKIP_IF_CURRENT_TEST_IS_DISABLED()
-        SetRefMode(LayerTestsUtils::IE);
+        SetRefMode(LayerTestsUtils::INTERPRETER);
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp:164]: I think, we can remove this line (It's default option)

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+        InferenceEngine::Precision netPrecision;
+        std::string targetDevice;
+        std::tie(should_decompose, seq_lenghts, batch, hidden_size, input_size, clip, ti_body, direction, netPrecision,
+                 targetDevice) = obj.param;
+        std::vector<std::vector<size_t>> inputShapes = {};
+
+        std::string type;
+        switch (ti_body) {
+            case TensorIteratorBody::LSTM:
+                type = "LSTM";
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:42]: We can move this logic to ngraph_helpers (Can be reused in the future, see overloaded operators in the end of file)

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        auto tensor_iterator = std::make_shared<ngraph::opset5::TensorIterator>();
+
+        // Each case consist of 3 steps:
+        // 1. Create TensorIterator body.
+        // 2. Set PortMap
+        // 3. Create outer function
+        auto axis = std::make_shared<ngraph::opset5::Constant>(ngraph::element::i64, ngraph::Shape{1}, std::vector<int64_t>{1});
+        switch (ti_body) {
+            case TensorIteratorBody::LSTM: {
+                inputShapes = {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:100]: Please, move ti creation logic to the [`builders`](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/ngraph_functions/include/ngraph_functions/builders.hpp)

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+                tensor_iterator->set_invariant_input(body_params[1], outer_params[1]);
+                tensor_iterator->get_iter_value(results[0]);
+
+                // 3. Outer function
+                function = std::make_shared<ngraph::Function>(ngraph::OutputVector{tensor_iterator->output(0), tensor_iterator->output(1)}, outer_params);
+                break;
+            }
+        }
+        if (should_decompose) {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:220]: We should do it in `SetUp()`? I offer to move it in test_body

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp [hunk 13]
```diff
@@ -53,7 +53,7 @@ namespace LayerTestsDefinitions {
 
     void LoopTest::SetUp() {
         SKIP_IF_CURRENT_TEST_IS_DISABLED()
-        SetRefMode(LayerTestsUtils::IE);
+        SetRefMode(LayerTestsUtils::INTERPRETER);
         bool execute_first_iteration;
         bool is_body_condition_const;
         bool body_condition; // works only if is_body_condition_const ==
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp:59]: I had asked you to move Loop creation to the builders from setup before (The same comment for TI). Could you please do it?

File: ngraph/core/reference/include/ngraph/runtime/reference/function.hpp (hunk 3 above)
Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: Sure, here majority implementations don't have return value:
https://github.com/openvinotoolkit/openvino/tree/master/ngraph/core/reference/include/ngraph/runtime/reference

[the remark this answers is not in the document]
Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp:30]: It is not a public API. And I think we can use `function` as a name because namespace `reference` should provide more information about purpose of this method.

**Review by ilyachur — APPROVED** [review]

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp (hunk 12 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:220]: It should be done after the creation of ngraph::function, I guess it a suitable place

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp (hunk 13 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp:59]: Not in this PR, have no time for it, not a blocker.

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp (hunk 11 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:100]: Not in this PR, have no time for it, not a blocker. Please create a separate ticket, if necessary

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp (hunk 10 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:42]: It should be done after the creation of ngraph::function, I guess it a suitable place

Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:42]: Not in this PR, have no time for it, not a blocker. Please create a separate ticket, if necessary

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp (hunk 12 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp:220]: `SetUp ` is called before the `ngraph::function` creation, I guess it's a suitable place for this code for now

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp (hunk 7 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp:27]: I test most of the TensorIterator capabilities in these tests.

File: ngraph/frontend/onnx_import/src/op/loop.cpp [hunk 14]
```diff
@@ -143,7 +143,8 @@ namespace ngraph
                     const auto concat_axis_const =
                         ngraph::op::Constant::create(ngraph::element::i64, {1}, {concat_axis});
                     // provide scalar handing for scan outputs
-                    for (int i = loop_carried_dependencies.size() + 1; i < body_outputs.size(); ++i)
+                    for (size_t i = loop_carried_dependencies.size() + 1; i < body_outputs.size();
+                         ++i)
```
[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/frontend/onnx_import/src/op/loop.cpp:147]: ngraph code style doesn't let me change this

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/loop.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/include/single_layer_tests/loop.hpp]: renamed function to PredefinedRefs, added new tests for it

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp (hunk 6 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/include/single_layer_tests/tensor_iterator.hpp]: done

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp (hunk 9 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/loop.cpp]: done

File: ngraph/core/reference/include/ngraph/runtime/reference/function.hpp (hunk 3 above)
Comment (itikhono) [reply ngraph/core/reference/include/ngraph/runtime/reference/function.hpp]: fixed

File: ngraph/core/reference/src/runtime/reference/function.cpp [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+#include "ngraph/runtime/reference/concat.hpp"
+#include "ngraph/runtime/tensor.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            bool call(const HostTensorVector& func_outputs,
```
[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/core/reference/src/runtime/reference/function.cpp]: fixed

File: ngraph/core/reference/include/ngraph/runtime/reference/loop.hpp [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            void loop(const std::shared_ptr<Function>& body,
+                      const op::util::OutputDescriptionVector& out_descs,
+                      const op::util::InputDescriptionVector& input_descs,
+                      const opset5::Loop::SpecialBodyPorts& special_ports,
+                      const HostTensorVector& out,
```
Comment (mikhail-treskin) [inline ngraph/core/reference/include/ngraph/runtime/reference/loop.hpp:32]: All other references operate with raw pointers instead of HostTensors which makes them more flexible. I propose to  change signatures of reference functions here and bellow (OK to do it in separate PR)

Comment (itikhono) [reply ngraph/core/reference/include/ngraph/runtime/reference/loop.hpp:32]: I agree. @mikhail-treskin Please create a new ticket in jira to not forget, and assign me

Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/loop.hpp:32]: Yeah, it is a good point. We already discussed it today with Ivan. But looks like it is require a lot of changes and we decided to do it later (we can do it later because it is not a public API).

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp (hunk 10 above)
Comment (iefode) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp]: Could you please create?

File: ngraph/core/reference/include/ngraph/runtime/reference/loop.hpp (hunk 16 above)
Comment (itikhono) [reply ngraph/core/reference/include/ngraph/runtime/reference/loop.hpp:32]: ticket 43220

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp (hunk 10 above)
Comment (itikhono) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/tensor_iterator.cpp]: 43220
